File: src/main/java/model/Album.java
package model;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class Album {
    private ArrayList<File> shuffledDirectory = new ArrayList<>();
    private int index = 0;

    public Album(String directory) throws Exception {
        File currentDirectory = new File(directory);
        File[] files = currentDirectory.listFiles(pathname -> {
            String extension = pathname.getName().toLowerCase();
            return extension.endsWith(".jpg") || extension.endsWith(".jpeg") || extension.endsWith(".png");
        });

        if(files == null) {
            throw new Exception("Given a bad directory. Couldn't find any files at: " + currentDirectory);
        }
        shuffledDirectory.addAll(Arrays.asList(files));
        Collections.shuffle(shuffledDirectory);
    }

    public File next(){
        if(index >= shuffledDirectory.size()) {
            Collections.shuffle(shuffledDirectory);
            index = 0;
        }
        return shuffledDirectory.get(index++);
    }

    public File previous() {
        index--;
        if(index < 0) {
            index = 0;
        }
        return shuffledDirectory.get(index);
    }
}

File: src/main/kotlin/model/Album.kt
package model

import java.io.File
import java.io.FilenameFilter
import java.util.*

class Album(directory: String?) {
    //private val shuffledDirectory = ArrayList<File?>()
    private val shuffledDirectory = ArrayList<String>()
    private var index = 0

    // Debugging memory issue
    private var fullCycle = 1

    init {
        val currentDirectory = File(directory)
        val files = currentDirectory.listFiles()
        val filesNames:MutableList<String> = mutableListOf()
        files?.map{
            if(validImage(it)) {
                filesNames.add(it!!.absolutePath)
            }
        }
        shuffledDirectory.addAll(filesNames)
        shuffledDirectory.shuffle()
    }

    private fun validImage(filename: File): Boolean {
        val extension = filename.name.lowercase(Locale.getDefault())
        return extension.endsWith(".jpg") || extension.endsWith(".jpeg") || extension.endsWith(".png")
    }

    operator fun next(): String {
        if (index >= shuffledDirectory.size) {
            shuffledDirectory.shuffle()
            index = 0
            println("Full cycles: ${fullCycle}, reshuffling images")
            fullCycle++
        }
        return shuffledDirectory[index++]
    }

    fun previous(): String {
        index--
        if (index < 0) {
            index = 0
        }
        return shuffledDirectory[index]
    }
}